### PR TITLE
Get rid of most generated variables

### LINKED ISSUE #1647: Get rid of most generated variables
Part of #1484.

Currently, we can't directly store a closure (a term and an environment together) into a Nickel AST. One reason is that we only have one AST, and a closure doesn't make any sense at e.g. parsing time, typechecking time and program transformation time. It would be strange to depend on runtime-related definitions for the core AST type.

But, during evaluation, we morally need to store closure (and in fact, thunks, for lazyness) inside expressions (typically, data structures such as arrays and records store thunks). The trick we use is to generate a unique variable, bind the closure in the environment, and store this variable in the term instead. This trick leads to a lot of variable generation and variable lookup, which go through the environment, meaning they incur hashmaps lookup (and, in the worst case, a linear-time lookup, because the environment is more or less a linked list of hashmaps). This is in particular a problem for the share normal form, which, for big terms (large records), might generate a huge load of heading lets binding generated variables, degrading the performance of even just parsing then exporting a Nickel program which is already in normal form (#1428).

### PR DESCRIPTION
**Pull request by yannham**
Fixes #1647.

## Context

This PR adds a new node to the AST, `Closure`, which contains a "pointer" (more abstractly, a cache index, in practice a thunk) to an allocated closure. This removes the need of closurizing terms in the environment everywhere.

The first motivation was performance, as we used to allocate specific variables and put them in the environment to store those closures, with renaming issues, environment propagations, and paying the price of environment insertion and lookup. Now, we basically just use direct pointers.

The price to pay is that `Closure` now leaks in the AST, even at stages where it doesn't make sense. Ideally, having separate ASTs for parsing and for evaluation would make this cleaner, and is the eventual plan - but for now, it's an ok price to pay to fix some big performance issues. Also, this change makes a lot of code much simpler, because we don't have to propagate and think about nested environments for closurization anymore. There's a lot less threading to do.

## Content

- Adds a `Closure` node to the AST
- Get rid of the share normal form transformation. This was the one creating a lot of let-binding with generated variables. Instead, we rely on the `closurized` attributes of arrays and records to know if something is a "fresh" array or record - in which case we allocate the closures for each element at first evaluation - or if it's already "transformed". In some sense, this PR factors the share normal form transformation directly into evaluation, removing the need of creating intermediate let-bindings and allocating the corresponding slot in the environment.
- Moves the trait `Closurizable` into its specific module, and rename it `Closurize`. We still use `closurize` because it's handy, but the signature is simpler.
- Updates various part of the codebase in consequence

## Effects

On the codebase provided for #1622, this change doesn't gain an order of magnitude but still seems to provide a boost of around 20-30% (it's hard to say more without a proper statistical method, because there is a lot of variance across runs - an effect of Sip hashing?). It's noticeable without being game changing.

However, on other examples, the gain is drastic. On an example reported in the online chat by another user, this:

```nickel
let things = std.array.range 0 3000
    |> std.array.map (fun number => {
        a = 0,
        b = 1,
        c = 2,
    })
in

things
    |> std.array.flat_map std.record.values
    |> std.array.at 0
```

Was basically taking forever (didn't finish before interrupted, and worked for at least a solid minute). After this change, it takes between 2 and 3 seconds (it's still a bit high, but nowhere near the previous numbers).

## Follow-up

One medium-term follow-up would be to split the AST, and have 1. a concrete AST with no mention of closures 2. A runtime AST with versions of `Array` and `(Rec)Record` which enforce their elements are stored inside closures.

### CODE REVIEW

File: core/src/term/mod.rs [hunk 1]
```diff
@@ -236,6 +237,89 @@ pub enum Term {
...
+
+    #[serde(skip)]
+    /// A "pointer" (cache index, which can see as a kind of generic pointer to the memory managed
+    /// by the evaluation cache) to a term together with its environment. Unfortunately, this is an
+    /// evaluation object leaking into the AST: ideally, we would have one concrete syntax tree
+    /// coming out of the parser, and a different representation for later stages, storing closures
+    /// and whatnot.
+    ///
+    /// This is not the case yet, so in the meantime, we have to mix everything together. The
+    /// ability to sotre closures directly in the AST without having to generate a variable a bind
```
Comment (jneem) [inline core/src/term/mod.rs:249]:
```suggestion
    /// ability to store closures directly in the AST without having to generate a variable and bind
```

File: core/src/term/mod.rs [hunk 2]
```diff
@@ -236,6 +237,89 @@ pub enum Term {
...
+            (Self::Annotated(l0, l1), Self::Annotated(r0, r1)) => l0 == r0 && l1 == r1,
+            (Self::Import(l0), Self::Import(r0)) => l0 == r0,
+            (Self::ResolvedImport(l0), Self::ResolvedImport(r0)) => l0 == r0,
+            (Self::Type(l0), Self::Type(r0)) => l0 == r0,
+            (Self::ParseError(l0), Self::ParseError(r0)) => l0 == r0,
+            (Self::RuntimeError(l0), Self::RuntimeError(r0)) => l0 == r0,
+            // We don't compare closure, because we can't, without the evaluation cache at hand.
+            // It's ok even if the cache index are the same: we implement PartialEq, so we can have
+            // `x != x`. In practice, this case shouldn't even be triggered, because tests usually
+            // compare simple terms without closures in it (or terms completely where closures have
```
Comment (jneem) [inline core/src/term/mod.rs:317]:
```suggestion
            // compare simple terms without closures in it (or terms where closures have
```

File: core/src/eval/fixpoint.rs [hunk 3]
```diff
@@ -54,21 +44,18 @@ fn patch_term<C: Cache>(
...
             if let Some(ref value) = field.value {
                 let idx = match value.as_ref() {
-                    Term::Var(ref var_id) => env
-                        .get(&var_id.ident())
-                        .cloned()
-                        .ok_or(EvalError::UnboundIdentifier(*var_id, value.pos))?,
+                    Term::Closure(idx) => idx.clone(),
                     _ => {
                         // If we are in this branch, `rt` must be a constant after the share normal
                         // form transformation, hence it should not need an environment, which is
```
Comment (jneem) [inline core/src/eval/fixpoint.rs:56]: What happens here now that the share normal form doesn't exist anymore?

**Review by jneem — APPROVED** [review]
I'm definitely lacking some context here, but I did have a look at least...

File: core/src/closurize.rs [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+        // ```
+        //
+        // Then, evaluating `foo` would unduly raise an unbound identifier error.
+        //
+        //
+        let pos = self.pos;
+
+        let idx = match_sharedterm! { self, with {
+                // We should always find a generated variable in the environment, but this method is
+                // not fallible, so we panic if we don't. We just wrap it in a new closure which will
```
[the remark this answers is not in the document]
Comment (yannham) [reply core/src/closurize.rs:105]: Good catch - in fact I think the whole comment has to go, or to be at least entirely rewritten to match the new situation.

File: core/src/eval/fixpoint.rs (hunk 3 above)
Comment (yannham) [reply core/src/eval/fixpoint.rs:56]: It is now enforced by the eval loop itself. I'll update this comment.

File: core/src/eval/fixpoint.rs [hunk 5]
```diff
@@ -81,22 +68,19 @@ pub fn rec_env<'a, I: Iterator<Item = (&'a LocIdent, &'a Field)>, C: Cache>(
...
+                    RichTerm::new(Term::Closure(idx), value.pos),
                     field.pending_contracts.iter().cloned(),
                     value.pos,
                 );
 
                 let final_closure = Closure {
                     body: with_ctr_applied,
-                    env: final_env,
+                    // FIXME: clean?
+                    env: Environment::new(),
```
[the remark this answers is not in the document]
Comment (yannham) [reply core/src/eval/fixpoint.rs:80]: Yeah, the comment needs to be reworked as well. What happens is that `final_env` was created here empty just above L84, anyway, and just served as a store for newly generated variables. So contracts and terms (unless they are constants, in which case we don't care) are actually already packed with their local environment, in the form of a `idx` (for all intent of purpose, this is a thunk). We just don't need this artificial `final_env` anymore, now that we have the `Closure` node

File: core/src/closurize.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yannham) [reply core/src/closurize.rs:105]: In fact no, your suggestion is the right patch. There are still generated variables around.

File: core/src/eval/fixpoint.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jneem) [reply core/src/eval/fixpoint.rs:78]: You wrote that `final_env` was created empty, but wasn't it a clone of the `env` parameter? Like, I understand how now we aren't generating variables and so we don't need to add any to the environment, but wasn't there an original record environment that needs to be passed on?

**Pull request merged**